Asynchro message2

robot269006 opened the pull request:
# WHAT
Configure asynchronous messaging function and incremental user search function

# WHY
For smoother message posting and incremental user search, by the use of asynchronous communication and ajax

robot269006 committed:
Create message API and preliminary test on console

1. Create js file and store under assets/javascripts director (No need
for gem ‘jquery-rails’ install, as it’s already there)
2. Set API controller
3. Test if input message shows up on console first in js

robot269006 committed:
Ajax settings

1. Set ajax type, data
2. Set success -> test with console.log
3. Set failure -> test with console.log
4. Set ajax url -> var postUrl = location.href;

*limit displayed messages to 5 as it messes up view
*small mistake noted in _group.haml so fixed it

robot269006 committed:
Ajax settings (set append function argument)

1. Set argument
2. Test with console.log

robot269006 committed:
Ajax settings (set append function, body only)

1. Append message input (no name or created date)
2. Test it out on console.log -> returns 1 object

robot269006 committed:
Ajax settings (set append function, name and date)

1. Set name and date
2. Delete console.log and append the return value to partial
3. Put ul class into var “list”, and append name, date, body to var
“list”
4. Limit the number of messages shown from 5 to 4 in index.haml

robot269006 committed:
Create search field

1.Create groups.js file under assets/javascript directory
2.Delete collection_box -> replace with search field, replace with
form_tag and text_field_tag

*not form_for, since it doesn’t create any values, and no models
associated

robot269006 committed:
Create js search function

1.Receive typed-in val() without event trigger (keyup)
2.Test by console.log -> whether or not the typed in info is reflected
real-time

robot269006 committed:
Set user source for search function

1. Create users_controller, and set search action (empty for now)
2. Set routing for users #search
3. rake routes
4. Write search action
 a. pass search result into @search instance
 b. @search result requested, request designated by ajax options
 c. js to show on console

robot269006 committed:
Set append function

1. Set appending function (buildHTML)
2. Set appending function to when successfully completed
3. Value retrieval fixed (trim)

ghost commented:
プルリクは細かくしたほうが達成感が得られやすいと思います。  
たとえば、今回の場合ですと、インクリメンタルサーチとメッセージ投稿の非同期化は別々のブランチにしたほうが、すぐにLGTMがでてやる気が出るかもしれません。  
私はこのように思いますが、どちらでもいいので参考までに。  
コミットのサイズが小さいとレビュワーがレビューしやすいというのもありますね。

tetsu8dayo commented on app/assets/javascripts/groups.js:
```diff
@@ -0,0 +1,38 @@
+    .done(function(data) {
+      var result;
+      data.forEach(function(data) {
+        console.log(data.name);
```
console.log は消しましょう

robot269006 replied:
失礼しました、削除します

robot269006 commented:
<img width="1265" alt="2017-05-19 19 08 16" src="https://cloud.githubusercontent.com/assets/24761206/26243361/90ec4a60-3cc6-11e7-87bb-75d7629e2bdb.png">

tetsu8dayo commented on app/assets/javascripts/groups.js:
```diff
@@ -0,0 +1,38 @@
+    }
+    preInput = input;
+  });
+});
```
検索機能はできていると思います。  
まだ検索したユーザをグループに追加する機能はできていないようですが、次のブランチになりますか？

robot269006 replied:
次のブランチで実施します。検索機能までなのかな、と思っていたのでこのプルリクではあげてません。一通りレビュー対応終わった後に実装します。

robot269006 commented:
プルリク今回一緒にあげてしまいすみませんでした。
修正対応の時間がまとまって取れなかったため一緒にあげてしまいました。
次回より気をつけます。。

tetsu8dayo commented on app/assets/javascripts/groups.js:
```diff
@@ -0,0 +1,38 @@
+$(window).on('load', function() {
+  function buildHTML(user) {
+    var html = '<div>' + user.name + '</div>'
```
この部分にxss の危険性があるとおもいます。確認していないのですが、ユーザーを保存する際にすでにエスケープがなされているので、このままで大丈夫なのかもしれません。しかし、推奨される書き方ではないと思うのでdomを構成するjavascript の機能を使うように置き換えましょう。  
このあたりは私は素人なので、ご自身で調べられることをお勧めいたします。  
以下のサイトが参考になると思います。  
https://railsguides.jp/security.html#%E3%82%AF%E3%83%AD%E3%82%B9%E3%82%B5%E3%82%A4%E3%83%88%E3%82%B9%E3%82%AF%E3%83%AA%E3%83%97%E3%83%86%E3%82%A3%E3%83%B3%E3%82%B0-xss

robot269006 replied:
参考文献までありがとうございます。確認し修正します

robot269006 replied:
確認しました。これで合ってるのかよくわかってませんが、結果、以下の修正をしました。
①$(document).readyに修正　→　ブラウザバージョンの対応
②ajaxのオプション内にcontentTypeの追記　→　json以外の通信を防ぐ
③インプットのサニタイズ　→　入力フォームへの悪意あるインプットを防ぐ

参考文献：
https://www.websec-room.com/2013/07/28/876
https://garafu.blogspot.jp/2013/12/javascript_15.html

tetsu8dayo commented on app/assets/javascripts/messages.js:
```diff
@@ -0,0 +1,32 @@
+    var date = list.append($('<li class="individualmessage__date">' + datareceiver.created_at + '<li>'));
+    var body = list.append($('<li class="individualmessage__body">' + datareceiver.body + '<li>'));
+    return list;
+  };
```
上と同様にxssの危険性があるような気がしますが、大丈夫でしょうか？

robot269006 replied:
確認します。

tetsu8dayo commented on app/controllers/messages_controller.rb:
```diff
@@ -4,15 +4,18 @@ class MessagesController < ApplicationController
     @groups = current_user.groups.includes(:users)
     @message = Message.new
-    @messages = Message.where(group_id: params[:group_id]).order('created_at ASC').includes(:user)
+    @messages = Message.where(group_id: params[:group_id]).limit(4).order('created_at DESC').includes(:user)
```
なぜlimit 4 なのでしょうか

robot269006 replied:
limit(5)だとビューがごちゃごちゃする（投稿したメッセージが投稿欄にかぶさる）のでlimit(4)にしてます。

ken-nag replied:
limitをかけると4件しかindexに表示できなくなってしまうのではないでしょうか

robot269006 replied:
はい、ビューが長くなってしまうためindexに最新のメッセージを４件だけ表示するようにしています。５件目がある状態（limit(5)）にした状態でメッセージ投稿すると、非同期通信で投稿したメッセージが投稿フォームにかぶさってしまうのでlimit(4)にしてます。

robot269006 replied:
<img width="1260" alt="2017-05-21 15 28 45" src="https://cloud.githubusercontent.com/assets/24761206/26281693/3075e00e-3e3a-11e7-919c-b2922389adfd.png">

robot269006 commented:
<img width="1261" alt="2017-05-19 19 17 06" src="https://cloud.githubusercontent.com/assets/24761206/26243734/dcf8b2da-3cc7-11e7-8087-0ed002f68003.png">
<img width="1256" alt="2017-05-19 19 17 24" src="https://cloud.githubusercontent.com/assets/24761206/26243735/dcfa8b32-3cc7-11e7-8b81-aa22ebaf54c1.png">

tetsu8dayo commented on config/routes.rb:
```diff
@@ -1,6 +1,11 @@
   root 'groups#index'
+  resources :users, only: [:search] do
+    collection do
+      get 'search', :action => :search
```
get  'search' だけで行けないでしょうか？　
resources を使って逆にコードが長くなっているので、使わなくて良いと思います。

robot269006 replied:
了解しました、ご指摘ありがとうございます！

robot269006 commented:
![message](https://cloud.githubusercontent.com/assets/24761206/26244065/3ba74fe8-3cc9-11e7-89f1-9cffcb5b0719.gif)
![user](https://cloud.githubusercontent.com/assets/24761206/26244066/3ba978b8-3cc9-11e7-838d-a7d44edf0191.gif)

ghost commented:
上のgif でundefined となっているところは、情報をうまく渡すことができていません。 undefined は javascript ではたしか初期化されていない変数や存在しないプロパティを参照した時に出るのだったと思います。（正確ではないかもしれませんが、初期化されていない変数を参照するというのは多分あっていると思います。

robot269006 committed:
Revision (re-route and XSS)

1. Delete console.log
2. Revise routes
3. Prevent XSS

robot269006 committed:
Revision (XSS)

1. Fixed window load -> document.ready
2. Deleted contentType from messages.js

robot269006 committed:
Revision (XSS, sanitize data)

added sanitizing codes to input area

robot269006 committed:
Revision (refactor sanitization into a function object)

robot269006 committed:
Revision (resolve undefined in groups.js)

Initialize variable “result” properly in groups.js -> resolve
“undefined” problem

robot269006 committed:
Revision (resolve undefined in messages.js)

1. Add hidden field that hands over user name in index.haml
2. Prepare a container for user name
3. Hand the user name container to variable “name” in buildHTML

robot269006 committed:
Revision (Small fixes, comments)

1. Sanitization refactoring
2. Add comments

robot269006 commented:
![message2](https://cloud.githubusercontent.com/assets/24761206/26281578/28e80848-3e36-11e7-8805-eecbab4a0120.gif)
![user2](https://cloud.githubusercontent.com/assets/24761206/26281579/28ee7dc2-3e36-11e7-9ba8-431c179cb437.gif)

robot269006 commented:
undefined問題も無事解決しました。レビューいただきありがとうございます。

ken-nag commented on app/controllers/messages_controller.rb:
```diff
@@ -4,15 +4,18 @@ class MessagesController < ApplicationController
-      redirect_to group_messages_path(@group), notice: "メッセージが作成されました"
+      respond_to do |format|
+        format.html { redirect_to group_messages_path(@group), notice: "メッセージが作成されました"}
+        format.json { render json: @message }
```
jbuilderを用いた実装にしましょう

robot269006 replied:
わかりました、調べて修正します。

ken-nag commented on app/assets/javascripts/messages.js:
```diff
@@ -0,0 +1,42 @@
+    var postUrl = location.href;
+    $.ajax({
+      type: 'POST',
+      url: postUrl + '.json',
```
dataTypeでjsonを指定しているのでここで.jsonを足す必要はありません

robot269006 replied:
ご指摘ありがとうございます、修正します。

robot269006 committed:
Revision (minor fix)

Delete ‘.json’ from url: option in ajax

robot269006 committed:
Revision (jbuilder)

1. Set rendering to jbuilder
2. Create jbuilder files
3. Delete nameField variable, and retrieve

robot269006 committed:
Revision (show all messages on index)

Add overflow properties

ken-nag commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,13 @@
+
+  def search
+    @users = User.where('name LIKE(?)', "%#{search_params[:keyword]}%").order('name ASC').limit(5)
+    render json: @users
```
ここでもjbuilderを使いましょう

robot269006 replied:
承知しました、対応します

robot269006 committed:
Revision (jbuilder for users)

Pass on user name by jbuilder

ken-nag commented:
変更点はokです
確認として
インクリメンタルサーチのgif画像と
上記のgif画像だとページの更新が行われていないかどうかわからないため、もう一度更新の有無がわかるようなmessage投稿のgif画像をお願いします

robot269006 committed:
Revision (fix users jbuilder)

it wasn’t giving the son data in array
1. Revise controller, ‘user/search’ -> ‘users/search’
2. Change name, /users.json.jbuilder -> /search.json.jbuilder
3. Output @users in an array by json.array!

robot269006 commented:
![message3](https://cloud.githubusercontent.com/assets/24761206/26282832/f4bbeaf8-3e53-11e7-950d-6114dd905a27.gif)
![users3](https://cloud.githubusercontent.com/assets/24761206/26282833/f4c3827c-3e53-11e7-818a-9a558ba2820d.gif)

ken-nag commented:
https://gyazo.com/a25e6d3af9b1b10ee1eb1acbb0da6d66
更新の有無を見るためにこの部分も写して送って頂きたいです

robot269006 commented:
![message4](https://cloud.githubusercontent.com/assets/24761206/26282877/3f6cd700-3e55-11e7-94df-841a9dbd466c.gif)
![user4](https://cloud.githubusercontent.com/assets/24761206/26282876/3f6b8058-3e55-11e7-9fa6-5230e8b26e23.gif)

ken-nag commented:
LGTM!!!!
